src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemorySegmentPool.java

```diff
@@ -0,0 +1,201 @@
+    // Rest 0
+  }
+
+  public MemorySegmentPool(ResourceScope scope) {
+    this(DEFAULT_MAX_SIZES, scope);
+  }
+  public MemorySegmentPool(int maxSizes[], ResourceScope scope) {
+    this.scope = scope;
+
+    validateMaxSizes(maxSizes);
+    for (int i=0; i < segmentsDequeue.length; i++) {
+      segmentsDequeue[i] = new SpinLockQueue<>(maxSizes[i]);
+    }
+  }
+
+  public SegmentAllocator allocatorForScope(ResourceScope resourceScope) {
+    // Prevent scope managing this pool to go away, when dependant allocator is alive
+    final var handle = scope.acquire();
+    resourceScope.addOnClose(handle::close);
+    return (bytesSize, bytesAlignment) -> getSegmentForScope(resourceScope, bytesSize, bytesAlignment);
+  }
+
+  public MemoryPoolSegment getSegmentEntryByLayout(MemoryLayout layout) {
+    return getSegmentEntryBySize(layout.byteSize(), layout.byteAlignment());
+  }
+
+  /**
+   * Gets segment from pool or allocates new one. Internally segments are cached.
+   * The size of segment can be larger than requested.
+   *
+   * @param size the size of segment.
+   *
+   * @return segment of size at least `size`
+   */
+  @ForceInline
+  public MemoryPoolSegment getSegmentEntryBySize(long size, long alignment) {
+    if (!scope.isAlive()) {
+      throw new IllegalStateException("Associated resource scope is closed");
+    }
+
+    final var alignedSize = (size + alignment - 1) & -alignment;
+    final var bitBound = bitBound(alignedSize);
+    final var segmentDequeue = segmentsDequeue[bitBound];
+
+    var segment = segmentDequeue.pollEntry();
+    if (segment == null) {
+      final var bitBoundedSize = 1L << bitBound;
+      segment = allocateNewEntry(segmentDequeue, bitBoundedSize);
+    }
+
+    return segment;
+  }
+
+  private MemorySegment getSegmentForScope(ResourceScope resourceScope, long size, long alignment) {
+    final var segmentEntry = getSegmentEntryBySize(size, alignment);
+    resourceScope.addOnClose(() -> segmentEntry.close());
+    return segmentEntry.memorySegment;
+  }
+
+  @ForceInline
+  private static int bitBound(long alignedSize) {
+    // If 100.., than 100... - 1 -> 01111
+    // If 101 -> than 101 - 1 -> 1....
+    return 64- Long.numberOfLeadingZeros(alignedSize - 1);
+  }
+
+  private MemoryPoolSegment allocateNewEntry(SpinLockQueue<MemoryPoolSegment> queue, long allocationSize) {
+    final var memoryAddress = CLinker.allocateMemory(allocationSize);
+    return new MemoryPoolSegment(queue,
+        (NativeMemorySegmentImpl) memoryAddress.asSegment(allocationSize, scope));
+  }
+
+  private static void validateMaxSizes(int maxSizes[]) {
+    Arrays.stream(maxSizes).filter(i -> i < 0).findAny()
+        .ifPresent(i -> {
+          throw new IllegalStateException("Invalid max size " + i);
+        });
+  }
+
+  public static class MemoryPoolSegment extends SpinLockQueue.Entry<MemoryPoolSegment> implements AutoCloseable {
+    private final NativeMemorySegmentImpl memorySegment;
```

Comment:
I think this should be MemoryAddress, and the scope should be "late bound" when a segment is retrieved from the queue and requested by a client. Otherwise clients get segments with a different scope than the one they are in - which in itself is ok, but I think it's counterintuitive for an allocator associated with a scope to return segments that last longer than that scope. 

In fact I'm not even 100% that you need a scope for the pool itself (which could mean that the acquire is not needed).

Reply:
I made a performance test, and it looks like creating new segment adds around 8ns to the test results (grow from 32 to 40).

I wonder if this would be a must have for direct pool API. Direct pool does not have to be associated with new scope.

However, for allocator version I think this should be done, as it can prevent leaking shared segment.

Reply:
It seems like I misunderstood some of what was going on in the patch. I think I was confused by the fact that memory didn't seem to be ever reclaimed - this led me to the conclusion that having a scope for the pool was redundant. But then, when inspecting a running application closer, I noted that segments are left in the pool - even if the scope is closed (in your original patch). That's because you use CLinker::allocate memory to allocate the segment, which is ok, but then there's never a cleanup action that goes from the pool scope closure to the closure of the segments in the pool. This is, I think, a bug.